File: 0022-Generate-Parentheses/solution.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        if n == 0:
            return [""]

        result = []
        current = []
        open_p = 0
        close_p = 0

        def dfs(result, current, open_p, close_p, n):
            if len(current) == 2*n:
                result.append("".join(current))
                return

            if open_p < n:
                current.append('(')
                open_p += 1
                dfs(result, current, open_p, close_p, n)
                open_p -= 1
                current.pop()

            if open_p > close_p:
                current.append(')')
                close_p += 1
                dfs(result, current, open_p, close_p, n)
                close_p -= 1
                current.pop()
                
        dfs(result, current, open_p, close_p, n)

        return result

File: 0022-Generate-Parentheses/solution.c
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
void dfs (char** result, int* returnSize, char* current, int open, 
int close, int n, int idx) {
    if (idx == 2 * n) {
        char *s = (char *)malloc((2 * n + 1) * sizeof(char));
        strcpy(s,current);
        result[*returnSize] = s;
        (*returnSize)++;
        return;
    }

    if (open < n) {
        current[idx] = '(';
        dfs(result, returnSize, current, open+1, close, n, idx + 1);
        
    }

    if (open > close) {
        current[idx] = ')';
        dfs(result, returnSize, current, open, close+1, n, idx+1);
    }
}


char** generateParenthesis(int n, int* returnSize) {
    *returnSize = 0;

    char** result = (char **)malloc(sizeof(char*) * 1500);
    char* current = (char *)malloc(sizeof(char) * (2 * n + 1));
    current[2 * n] = '\0';

    dfs (result, returnSize, current, 0, 0, n, 0); 

    free(current);
    return result;
}
